cps/kni.c

```diff
@@ -0,0 +1,242 @@
+		RTE_LOG(ERR, GATEKEEPER, "cps: open: %s\n", strerror(errno));
+		return NULL;
+	}
+
+	ret = fstat(fd, &stat_buf);
+	if (ret < 0) {
+		RTE_LOG(ERR, GATEKEEPER, "cps: fstat: %s\n", strerror(errno));
+		goto close;
+	}
+
+	kmod_size = stat_buf.st_size;
+
+	buffer = rte_malloc("kni_kmod", kmod_size, 0);
+	if (buffer == NULL) {
+		RTE_LOG(ERR, GATEKEEPER,
+			"cps: couldn't allocate %u bytes to read %s\n",
+			kmod_size, filename);
+		goto close;
+	}
+
+	*size = 0;
+	while ((ret = read(fd, buffer + *size, kmod_size - *size)) > 0)
+		*size += ret;
+	if (ret < 0)
+		goto free;
+
+	RTE_VERIFY(*size == kmod_size);
+	close(fd);
+	return buffer;
+
+free:
+	rte_free(buffer);
+close:
+	close(fd);
+	return NULL;
+}
+
+static const char *
+moderror(int err)
+{
+	switch (err) {
+	case ENOEXEC:
+		return "Invalid module format";
+	case ENOENT:
+		return "Unknown symbol in module";
+	case ESRCH:
+		return "Module has wrong symbol version";
+	case EINVAL:
+		return "Invalid parameters";
+	default:
+		return strerror(err);
+	}
+}
+
+int
+init_kni(const char *kni_kmod_path, unsigned int num_kni)
+{
+	unsigned long len;
+	int ret;
+
+	void *file = grab_file(kni_kmod_path, &len);
+	if (file == NULL) {
+		RTE_LOG(ERR, GATEKEEPER, "insmod: can't read '%s'\n",
+			kni_kmod_path);
+		return -1;
+	}
+
+	ret = init_module(file, len, "");
+	if (ret < 0) {
+		RTE_LOG(ERR, GATEKEEPER,
+			"insmod: error inserting '%s': %d %s\n",
+			kni_kmod_path, ret, moderror(errno));
+		rte_free(file);
+		return ret;
+	}
+	rte_free(file);
+
+	rte_kni_init(num_kni);
+	return 0;
+}
+
+static int
+check_usage(const char *modname)
+{
+	FILE *module_list;
+	int found_mods = false;
+	char line[10240], name[64];
+	int ret = 0;
+
+	module_list = fopen("/proc/modules", "r");
+	if (module_list == NULL) {
+		RTE_LOG(ERR, GATEKEEPER,
+			"Can't open /proc/modules: %s\n", strerror(errno));
+		return -1;
+	}
+
+	while (fgets(line, sizeof(line), module_list) != NULL) {
+		unsigned long size, refs;
+		int scanned;
+
+		found_mods = true;
+
+		if (strchr(line, '\n') == NULL) {
+			RTE_LOG(ERR, GATEKEEPER, "long line broke rmmod.\n");
+			ret = -1;
+			goto out;
+		}
+
+		/*
+		 * Bound number of bytes written to @name; maximum
+		 * field width does not include null terminator.
+		 */
+		scanned = sscanf(line, "%63s %lu %lu", name, &size, &refs);
+
+		if (scanned <= 2 || scanned == EOF) {
+			if (scanned < 2 || scanned == EOF)
+				RTE_LOG(ERR, GATEKEEPER,
+					"Unknown format in /proc/modules: %s\n",
+					line);
+			else
+				RTE_LOG(ERR, GATEKEEPER,
+					"Kernel doesn't support unloading.\n");
+			ret = -1;
+			goto out;
+		}
+
+		if (strcmp(name, modname) != 0)
+			continue;
+
+		if (refs != 0) {
+			RTE_LOG(ERR, GATEKEEPER,
+				"Module %s is in use\n", modname);
+			ret = -1;
+		}
+
+		goto out;
+	}
+
+	if (found_mods)
+		RTE_LOG(ERR, GATEKEEPER,
+			"Module %s does not exist in /proc/modules\n", modname);
+	else
+		RTE_LOG(ERR, GATEKEEPER,
+			"fgets: error in reading %s\n", modname);
```

Comment:
1. Shouldn't the format string be `"fgets: error in reading /proc/modules\n"`?

2. Add define `PROC_MODULES_FILENAME` just before `check_usage()` to avoid repeating `/proc/modules` three times.